apps/dashboard/src/app/(dashboard)/dashboard/connect/ecosystem/[slug]/(active)/components/client/update-partner-form.client.tsx

```diff
@@ -0,0 +1,249 @@
+import { ToolTipLabel } from "../../../../../../../../../@/components/ui/tooltip";
+import type { Ecosystem, Partner } from "../../../../types";
+import { isDomainRegex } from "../../constants";
+import { useUpdatePartner } from "../../hooks/use-update-partner";
+
+const formSchema = z.object({
+  name: z
+    .string()
+    .trim()
+    .min(3, {
+      message: "Name must be at least 3 characters",
+    })
+    .refine((name) => /^[a-zA-Z0-9 ]*$/.test(name), {
+      message: "Name can only contain letters, numbers and spaces",
+    }),
+  domains: z
+    .string()
+    .trim()
+    .transform((s) => s.split(/,| /).filter((d) => d.length > 0))
+    .refine((domains) => domains.every((d) => isDomainRegex.test(d)), {
+      message: "Invalid domain format", // This error message CANNOT be within the array iteration, or the FormMessage won't be able to find it in the form state
+    })
+    .transform((s) => s.join(",")), // This is rejoined to return a string (and later split again) since react-hook-form's typings can't hnadle different input vs output types
+  bundleIds: z
+    .string()
+    .trim()
+    .transform((s) =>
+      s
+        .split(/,| /)
+        .filter((d) => d.length > 0)
+        .join(","),
+    ),
+  permissions: z
+    .string({
+      required_error: "Please specify wallet permissions",
+    }) // We use a string rather than an enum here so the defualt value will work
+    .refine((p) => ["PROMPT_USER_V1", "FULL_CONTROL_V1"].includes(p), {
+      message: "Invalid permissions setting",
+    }),
+});
+
+export function UpdatePartnerForm({
+  ecosystem,
+  partner,
+  onSuccess,
+}: { ecosystem: Ecosystem; partner: Partner; onSuccess: () => void }) {
+  const form = useForm<z.input<typeof formSchema>>({
+    resolver: zodResolver(formSchema),
+    defaultValues: {
+      name: partner.name,
+      domains: partner.allowlistedDomains.join(","),
+      bundleIds: partner.allowlistedBundleIds.join(","),
+      permissions: partner.permissions[0],
+    },
+  });
+
+  const { updatePartner, isLoading } = useUpdatePartner({
+    onSuccess: () => {
+      form.reset();
+      onSuccess();
+    },
+    onError: (error) => {
+      const message =
+        error instanceof Error
+          ? error.message
+          : "Failed to add ecosystem partner";
+      toast.error(message);
+    },
+  });
+
+  return (
+    <Form {...form}>
+      <form
+        onSubmit={form.handleSubmit((values) => {
+          updatePartner({
+            ecosystem,
+            partnerId: partner.id,
+            name: values.name,
+            allowlistedDomains: values.domains
+              .split(/,| /)
+              .filter((d) => d.length > 0),
+            allowlistedBundleIds: values.bundleIds
+              .split(/,| /)
+              .filter((d) => d.length > 0),
+            permissions: [
+              values.permissions as unknown as
+                | "FULL_CONTROL_V1"
+                | "PROMPT_USER_V1",
+            ],
```

Comment:
is permissions a string or an array? why does it take an array if it's a "switch"?

Reply:
It's an array behind the scenes / in the backend, I think there's opportunity for there to be multiple permissions on one partner in the future

Reply:
okay so here what we're doing is coercing the form (string/enum) into the array?

Reply:
Correct